Exposure Parsing

z3z1ma opened the pull request:
# Description

Reposting because some branch renaming fudged up the PR. This is still ready for review. 

Parsing exposures from Metabase automatically in relation to a dbt project. This function is huge for understanding exposures within your data model. Close the loop between model composition and tangible BI exposures of those models creating an extremely valuable company asset in the dbt docs. In the nature of this project, the aim is we can use almost the same invocation parameters users set up in CI / CD for `dbt-metabase export` as they can with `dbt-metabase exposures`. This lets you run the invocation in production to essentially keep you docs in full sync with BI.

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?

Open to ideas but ultimately I think making a mock api from metabase running on top of a database that has been seeded and ran through with the dbt jaffle shop project would be most robust. Just a lot of work. 

- [x] CI test for exposure documentation

**Test Configuration**:
* Python version: 3.8

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

References #22

z3z1ma commented:
@gouline 

This PR is ready to rock,
I particularly think you will like the unit test structure (full mock api) as well as utility suite for recreation I have put together for the mb client. But otherwise we are good.

gouline commented on dbtmetabase/__init__.py:
```diff
@@ -213,6 +241,29 @@ def main(args: List = None):
         default=False,
         help="Append tags to Table descriptions in Metabase",
     )
+    parser.add_argument(
```
Arguments that only apply to exposures should go in a subparser. Here's an example https://www.reddit.com/r/learnpython/comments/7xj2f2/correct_way_to_make_nested_argparse/du8qcs2

z3z1ma replied:
> Arguments that only apply to exposures should go in a subparser. Here's an example https://www.reddit.com/r/learnpython/comments/7xj2f2/correct_way_to_make_nested_argparse/du8qcs2

I opted for arg groups because when I set up subparsers, it split up the help text which seemed suboptimal when there are only a small group of args specific to the function. I am still open to full on sub parser though so I'll leave this comment open until you weigh in further. (I have a pic in the PR conversation of the grouping)

gouline commented on dbtmetabase/metabase.py:
```diff
@@ -139,7 +164,13 @@ def models_compatible(self, database_id: str, models: Sequence) -> bool:
+        database: str,
+        models: Sequence[MetabaseModel],
+        aliases,
+        **kwargs,
```
This is why I want to move to a `Config` object, kwargs everywhere will be a pain to debug and click through in people's IDEs when they're figuring out how it works.

z3z1ma replied:
Agreed with the config obj, will be a much cleaner method

gouline commented on dbtmetabase/metabase.py:
```diff
@@ -37,6 +50,15 @@ def __init__(
+        self.tables: Iterable = []
+        self.table_map: MutableMapping = {}
+        self.models_exposed: List = []
+        self.mb_native_query: str = ""
```
Don't use the `mb_` prefix. Externally, I moved to `metabase_`, but here you're already in the Metabase context, so no reason for the prefix altogether.

z3z1ma replied:
removed prefix, thanks!

gouline commented on dbtmetabase/__init__.py:
```diff
@@ -227,23 +278,41 @@ def main(args: List = None):
+        )
+    else:
+        logging.warning(
+            "Invalid command. Must be one of either 'export' or 'exposures'."
```
Would you say it "exports" the exposures as well or am I misunderstanding what this is doing? If so, maybe we have `export-models` and `export-exposures` or even `models` and `exposures`?

z3z1ma replied:
Hm well since we called export being to movement of information from dbt -> metabase,
this is essentially the inverse; so if speaking from the perspective of the dbt project-
We would export model defs to metabase
We would import metabase exposures to dbt project

But to save for unneeded complexity, I like the idea of just `models` and `exposures` (and eventually `metrics` to round everything out). Very simple that way which is good for users.

gouline commented on dbtmetabase/metabase.py:
```diff
@@ -403,14 +436,198 @@ def build_metadata_lookups(
+            if collection["name"] in exclude_collections:
+                continue
+            if not include_personal_collections and collection["name"].endswith(
+                "'s Personal Collection"
```
You don't need to parse the name to check that it's a personal collection, `personal_owner_id` will be non-null (or null for non-personal collections).

z3z1ma replied:
Adjusted to use key, cleaner simpler way to do it for sure since we have it

gouline commented on tests/fixtures/exposure/baseline_test_exposures.yml:
```diff
@@ -0,0 +1,398 @@
+  maturity: medium
+  owner:
+    name: Alexander Butler
+    email: alex@source.co
```
Are you comfortable making your email address visible and searchable? Maybe replace with something ending in `@example.com` or something.

z3z1ma replied:
oops good call, thanks for catching that

gouline reviewed (changes requested):
Massive effort, well done!

This is my first-pass review that focuses on structure of the programmatic and CLI API, which we need to carefully consider as this tool grows to ensure it's still easy and not intimidating to use.

z3z1ma commented on dbtmetabase/__init__.py:
```diff
@@ -1,35 +1,49 @@
 
-def export(
-    dbt_database: str,
+def execute(
```
Sounds genius actually to make usage of the package super slick, how would you envision this?
Would the first invocation cache values into a yaml if it does not already exist? And passed flags update the yaml if there is a diff and invocation succeeds?

z3z1ma replied:
But inadvertently pooling the two funcs to one made the list bigger so there ya go ha

z3z1ma replied:
config object model completed

z3z1ma commented on README.rst:
```diff
@@ -110,6 +112,59 @@ Check your Metabase instance by going into Settings > Admin > Data Model, you
+        depends_on:
+        - ref('orders')
+
+Questions which are Native queries will have the SQL propagate to a code block in the documentation
```
adjusted, thanks

z3z1ma commented on README.rst:
```diff
@@ -110,6 +112,59 @@ Check your Metabase instance by going into Settings > Admin > Data Model, you
+        url: http://your.metabase/card/8
+        maturity: medium
+        owner:
+          name: Jeff Bezos
```
Moved to proper fictional

z3z1ma commented on README.rst:
```diff
@@ -110,6 +112,59 @@ Check your Metabase instance by going into Settings > Admin > Data Model, you
+        --metabase_user user@example.com \
+        --metabase_password Password123 \
+        --metabase_database business \
+        --dbt_docs_url https://your.metabase/ \
```
right the arg actually isnt prevalent for exposures- removed, thank you

z3z1ma commented on README.rst:
```diff
@@ -260,10 +315,10 @@ line. But if you prefer to call it from your code, here's how to do it:
 
-    dbtmetabase.export(
+    dbtmetabase.execute(
+      command="export_models",
```
corrected with interface update

z3z1ma commented:
@gouline 

I think we have addressed the concerns now well.
the cli and program interface is greatly improved as well as programmatic through the config objs and arg groups

![image](https://user-images.githubusercontent.com/41213451/126611334-7993240b-f439-42a6-8844-7e3c05837d6c.png)


```
dbt-metabase models --dbt_database test --dbt_manifest_path tests/fixtures/sample_project/target/manifest.json --metabase_host localhost:3000 --metabase_user alex@... --metabase_password "..." --metabase_database unit_testing --metabase_use_http
```

```
dbt-metabase exposures --dbt_database test --dbt_manifest_path tests/fixtures/sample_project/target/manifest.json --metabase_host localhost:3000 --metabase_user alex@... --metabase_password "..." --metabase_database unit_testing --metabase_use_http
```

Example tested commands above show how easy it is to use one command or another. The latter command defaults the output yaml name to `metabase_exposures.yml` and the output path to the local dir. Both of those could be provided as args but for the sake of the example- the commands with minimal args can be used interchangeably. 


I am racking my brain on some pylinting issue now that was not occuring before

************* Module dbtmetabase.parsers.dbt_manifest
dbtmetabase/parsers/dbt_manifest.py:6:0: E0611: No name 'metabase' in module 'models' (no-name-in-module)
dbtmetabase/parsers/dbt_manifest.py:7:0: E0611: No name 'metabase' in module 'models' (no-name-in-module)
************* Module dbtmetabase.parsers.dbt_folder
dbtmetabase/parsers/dbt_folder.py:10:0: E0611: No name 'metabase' in module 'models' (no-name-in-module)
dbtmetabase/parsers/dbt_folder.py:11:0: E0611: No name 'metabase' in module 'models' (no-name-in-module)

erika-e commented:
@z3z1ma I had a go at testing this with my local setup. I ran into two problems. They may be errors related to how I got setup to test, so I'm open to feedback on a better way to go about testing! I'm running metabase in docker, connected to my development schema and using a `manifest.json` generated from a development run. I have Python 3.9.5 in the venv I'm using for testing.

I installed this branch to the venv I've been using for testing with `pip3 install git+https://github.com/z3z1ma/dbt-metabase.git@exposures`

The output of pip freeze now includes `dbt-metabase @ git+https://github.com/z3z1ma/dbt-metabase.git@32f409c83ae6af77963f25aba573c6b0334991ee`

Initially trying to run dbt-metabase I got an error -- will leave another comment about that, since I found a workaround for testing purposes.

I can run `dbt-metabase models`  I had to add a `--schema_excludes` argument to avoid errors which I didn't need previously.

I get an error when I try to run `dbt-metabase exposures` :

```
2021-07-23 11:07:08,763 - INFO - Session established successfully
Traceback (most recent call last):
  File "/Users/erikapullum/metabase-python/metabase-env/bin/dbt-metabase", line 6, in <module>
    dbtmetabase.main(sys.argv[1:])
  File "/Users/erikapullum/metabase-python/metabase-env/lib/python3.9/site-packages/dbtmetabase/__init__.py", line 345, in main
    exposures(
  File "/Users/erikapullum/metabase-python/metabase-env/lib/python3.9/site-packages/dbtmetabase/__init__.py", line 142, in exposures
    dbt_models = reader.read_models(
UnboundLocalError: local variable 'reader' referenced before assignment
```

erika-e commented on dbtmetabase/__init__.py:
```diff
@@ -6,112 +6,165 @@
-from typing import Iterable, List, Union
+from typing import Iterable, List, Union, Optional
 
 from ._version import version as __version__
```
I get an error from this line. May be related to my setup -- see [other comment](https://github.com/gouline/dbt-metabase/pull/38#issuecomment-885706487) here.

If I comment this line out I can run `dbt-metabase models` with no errors.

z3z1ma replied:
@gouline 

I also get an error on this line. Wondering if there is a step we are missing in the makefile / setup? 
I did a `pip install -e .` in the local dir which of course finds a setup py. I see in the toml we should be generating it. Just not sure why mine hasnt generated when installing either via `git clone`, `cd`, `pip install -e .` or with `pip install git+https://github.com/z3z1ma/dbt-metabase.git@exposures`

Maybe just something I am missing? Ill test it again in another fresh venv.

gouline replied:
I will need to document this, but `_version.py` is generated dynamically to allow automatic releases from GitHub Actions. It's not really intended to be installed from Git, you have to run `make` (or `make build`) and if you want to install locally, there's `make dev-install` for that.

However, if you have that use case, I'll think about either including a dummy `_version.py` or making that import optional. I'll experiment and see what works 🤔

gouline replied:
The version issue should now be fixed in master (see #42). You should be able to install using pip from a git branch, like @z3z1ma wrote above.

erika-e commented:
> ```
> UnboundLocalError: local variable 'reader' referenced before assignment
> ```

Resolved by specifying the `--dbt_manifest_path`. I did not get an error message when running with no path argument specified.

z3z1ma commented:
@erika-e 

Thanks so much for your time so far! 

Here are some minimal working examples I just ran now from CLI I personally use for testing on docker containers (I also use my branches in production containers personally being an avid stress tester)

```
dbt-metabase models --dbt_database test --dbt_manifest_path tests/fixtures/sample_project/target/manifest.json --metabase_host localhost:3000 --metabase_user alex@... --metabase_password "..." --metabase_database unit_testing --metabase_use_http

dbt-metabase exposures --dbt_database test **--dbt_path tests/fixtures/sample_project** --metabase_host localhost:3000 --metabase_user alex@... --metabase_password ... --metabase_database unit_testing --metabase_use_http

dbt-metabase exposures --dbt_database test **--dbt_manifest_path tests/fixtures/sample_project/target/manifest.json** --metabase_host localhost:3000 --metabase_user alex@... --metabase_password "..." --metabase_database unit_testing --metabase_use_http
```

Can we explore why your invocation did not work or what specific flags you supplied? Definitely sounds like an assertion issue or some combination of flags that shouldve supplied you an informative alert. Schema excludes shouldnt ever be required either.

erika-e commented:
@z3z1ma Sure thing! Here are the invocations: 

<details><summary> dbt-metabase models invocation</summary>

```
dbt-metabase models \
--dbt_manifest_path dbt_manifest_semantic_types/manifest.json \
--dbt_database our_database \
--metabase_host localhost:3000 \
--metabase_user eep@... \
--metabase_password '...' \
--metabase_database EPullum_Dev \
--metabase_use_http \
--schema DEV_EPULLUM \
--schema_excludes 'DEV_EPULLUM_STG' 'DEV_EPULLUM_EX'
```

Without the `--schema_excludes` I get errors for tables in the excluded schemas that are not found.

</details>

What is the purpose of the `**` around the `--dbt_manifest_path` argument? That's the first difference that jumps out.

<details><summary> dbt-metabase exposures invocation </summary>

```
dbt-metabase exposures \
--dbt_manifest_path dbt_manifest_semantic_types/manifest.json \
--dbt_database ANALYTICS \
--metabase_host localhost:3000 \
--metabase_user eep@... \
--metabase_password '...' \
--metabase_database EPullum_Dev \
--metabase_use_http \
--schema DEV_EPULLUM 
```
</details>

I am able to run `dbt-metabase exposures` command and I get a message out that says (for example)
```
2021-07-23 12:13:11,014 - INFO - Introspecting card: Web and Mobile Event Counts
2021-07-23 12:13:11,015 - INFO - Model extracted from Metabase question: DAILY_ACTIVITY
```
However, the `depends_on` section of the `exposures.yml` file is empty.

<details><summary>metabase_exposures.yml example</summary>

```
- name: Web_and_Mobile_Event_Counts
  description: '### Visualization: Line


    Event counts by day


    #### Metadata


    Metabase Id: __5__


    Created On: __2021-07-23T15:48:34.42767Z__'
  type: analysis
  url: http://localhost:3000/card/5
  maturity: medium
  owner:
    name: eep
    email: eep@...
  depends_on: []
```

</details>

z3z1ma commented:
> What is the purpose of the ** around the --dbt_manifest_path argument? That's the first difference that jumps out.

@erika-e  Sorry that was just a markdown fail, just meant to emphasize that `exposures` worked with _either_ folder parser or manifest and did not require manifest specifically.

So -depends_on can only be empty if there is an orphaned exposure in the context of the run (ie no refable model from the dbt parser found matching metabase table name). So another good addition would be the inclusion of an orphaned exposure logging message rather than proceeding quietly. 

As far as how or why it happened here, the first place I would look is to see that the question is built on your expected schema `DEV_EPULLUM `, seems silly but always best to start from the bottom up.

erika-e commented:
> So -depends_on can only be empty if there is an orphaned exposure in the context of the run (ie no refable model from the dbt parser found matching metabase table name). So another good addition would be the inclusion of an orphaned exposure logging message rather than proceeding quietly.
> 
> As far as how or why it happened here, the first place I would look is to see that the question is built on your expected schema `DEV_EPULLUM `, seems silly but always best to start from the bottom up.

Let me make sure I understand a couple of things:
`--schema` I understood to be referring to the schema within the database where the models are being built - in this case for me `DEV_EPULLUM` - and this argument pertains to dbt and Metabase?

`--metabase_database` is the name of the database within Metabase. The help text says "Target database name as set in Metabase (typically aliased)"

The name of the database in metabase (on the settings page) is `EPullum_Dev` and the schema it's connected to is `DEV_EPULLUM`. In my test instance that's the only database the question could be pulling that table from.

Depends on is blank for all questions/dashboards. I made one that is just select * from test_table, and depends_on is also blank for it. So it does seem like some kind of invocation problem.

z3z1ma commented:
@erika-e 

Both `schema` and `schema_excludes` only restrict the scope of the dbt parser.
the same follows for the `includes` and `excludes` arguments (so for example with `--includes my_model` appended to your typical invocaiton, you can propogate docs / extract exposures to an output file for a single model. 

Everything is pulled from Metabase and checked against the subset of dbt models for table (model) name matches from dbt <-> Metabase. 

test_table is a model in your manifest/dbt project?

erika-e commented on dbtmetabase/__init__.py:
```diff
@@ -6,112 +6,165 @@
-        metabase_verify (Union[str, bool], optional): Supply path to certificate or disable verification. Defaults to True.
-        metabase_sync_skip (bool, optional): Skip synchronizing Metabase database before export. Defaults to False.
-        metabase_sync_timeout (int, optional): Metabase synchronization timeout in seconds. Defaults to None.
-        schema (str, optional): Target schema name. Defaults to "public".
```
For consistency, should all arguments specific to dbt be prefixed with `dbt_`? Per @z3z1ma in [this comment](https://github.com/gouline/dbt-metabase/pull/38#issuecomment-885781332) schema for ex applies to the schemas from dbt. 

-  `schema` -> `dbt_schema`
- `exclude_schema` -> `dbt_exclude_schema`
- `includes` -> `dbt_include_models`
- `excludes` -> `dbt_exclude_models`

Those names are longer but also convey more context about their purpose.

z3z1ma replied:
With the addition of the new command, we are working in CLI improvements in this PR. So I'm totally open to the idea of making those specific flags more clear @erika-e. `schema` used to be used in the context of Metabase but with the introduction of the hash key of schema.table being used for Metabase references, it became unnecessary to filter the Metabase table and field lookups. Instead filtering the dbt project for a subset of models was sufficient to scope the execution. 

I wanted to avoid the breaking change flag on this one but if it's an improvement for potential users and existing users I think @goulline would agree.

gouline replied:
Don't worry about breaking changes, we already made _a lot_ of breaking changes.

erika-e replied:
I could take on making the arguement name changes in a separate PR, since there's a lot going on with this one already. @gouline would you prefer that we make issues for things like this? I'm happy to make one if you like.

gouline replied:
I'd say yes for something large, but for renaming arguments feel free to just do it and reference this discussion thread in your PR description.

erika-e commented:
> @erika-e
> 
> Both `schema` and `schema_excludes` only restrict the scope of the dbt parser.
> the same follows for the `includes` and `excludes` arguments (so for example with `--includes my_model` appended to your typical invocation, you can propogate docs / extract exposures to an output file for a single model.
> 
> Everything is pulled from Metabase and checked against the subset of dbt models for table (model) name matches from dbt <-> Metabase.
> 
> test_table is a model in your manifest/dbt project?

Yes, I changed the name from the real one to the more generic `test_table` but it's definitely there. I checked the manifest to be sure and it's there with the same name. Totally possible it's user error still. 😅 I'll do some more testing and debugging and see if I can pin the problem down.

erika-e commented on dbtmetabase/__init__.py:
```diff
@@ -120,99 +173,133 @@ def main(args: List = None):
+    parser_models = parser.add_argument_group("`models` optional arguments")
+    parser_exposures = parser.add_argument_group("`exposures` optional arguments")
+
+    # dbt arguments
```
@z3z1ma It looks like `--schema` was dropped from this list of arguments. Since it has the default of public there's not an error message. I think this in combination with [how I was invoking](https://github.com/gouline/dbt-metabase/pull/38#issuecomment-885759913) is the root cause of the missing exposures.

erika-e commented on dbtmetabase/__init__.py:
```diff
@@ -6,112 +6,165 @@
+        dbt_config.schema_excludes = {
+            schema.upper() for schema in dbt_config.schema_excludes
+        }
+
```
To troubleshoot I added code here to print out dbt_config.schema and dbt_config.schema_excludes. 

With this invocation:
dbt-metabase exposures \
--dbt_manifest_path dbt_manifest_semantic_types/manifest.json \
--dbt_database ANALYTICS \
--metabase_host localhost:3000 \
--metabase_user eep@... \
--metabase_password '...' \
--metabase_database EPullum_Dev \
--metabase_use_http \
--schema DEV_EPULLUM 

dbt_config.schema has a value of None
dbt_config.schema_excludes has a value of {'DEV_EPULLUM'}

z3z1ma replied:
@erika-e 

Thanks for your diligence, you rock.
Your totally right. I am managing two PRs right now and I during my own code review processes would've caught and fixed the issue but only applied the change to my latest branch `metrics` 

I will add a commit today that addresses.

erika-e replied:
Happy to help! 😃   Exposures parsing is going to be an awesome tool for us so I'm glad to be able to support and contribute.

z3z1ma commented:
@gouline 
All changes based on excellent user usage review as well as my own usage has been implemented. Now I need to integrate any readme updates to point out the renamed flags, but functionally, I dont think there is much more to do. The unit tests are very good in displaying the function on a full mock api with dbt model project seeded generating the exact expectation of auto generated exposure documentation.

gouline commented on dbtmetabase/models/config.py:
```diff
@@ -0,0 +1,30 @@
+
+
+@dataclass
+class dbtConfig:
```
This should be `DbtConfig`. I know dbt themselves sometimes use lower case, but that's against Python style guidelines and we already have `DbtFolderReader` etc, so I prefer `DbtConfig`.

z3z1ma replied:
Changed to `DbtConfig` to conform to style guidlines

gouline commented on dbtmetabase/parsers/dbt_folder.py:
```diff
@@ -29,12 +29,12 @@ def __init__(self, project_path: str):
         excludes: Iterable = None,
         include_tags: bool = True,
-        docs_url: str = None,
+        dbt_docs_url: str = None,
```
Prefix `dbt_` makes sense for entry point options, but here it's going into dbt-only context, so I think `docs_url` is enough.

z3z1ma replied:
agreed, adjusted accordingly

gouline commented on dbtmetabase/parsers/dbt_folder.py:
```diff
@@ -52,17 +52,19 @@ def read_models(
             )
 
-        if docs_url:
+        if dbt_docs_url:
```
I know this is existing code, but I missed it in the last review. This logger should be in the `export()` method, not the parser itself. The user may have gotten here via programmatic API, so the CLI flag would be meaningless to them.

z3z1ma replied:
It was originally for passing lint checks since args have to be used but we want the API for both readers to be the same without using kwargs. But good point, I have moved this to warn in the init and adjusted the code to pass linting still

gouline reviewed (changes requested):
Looking good, only a few small issues left. Almost ready for merging.

z3z1ma commented:
@gouline 

All requested changes complete. Plus a little cherry in ensuring our config objects input args match our cli flag args verbatim for additional ease of use between the two methods of utilizing the package.

gouline commented on dbtmetabase/models/config.py:
```diff
@@ -0,0 +1,30 @@
+
+
+@dataclass
+class MetabaseConfig:
```
Likely last thing, but what's the purpose of essentially double prefixing by having `metabase_` for every field in an object that's already prefixed `Metabase`? Same with `dbt_` and `Dbt` below. Can we drop these field prefixes?

z3z1ma replied:
agreed, done

gouline commented on dbtmetabase/metabase.py:
```diff
@@ -404,14 +434,192 @@ def build_metadata_lookups(
+        duplicate_name_check = []
+        captured_exposures = []
+
+        for collection in self.collections:
```
This loop is very long and I'm struggling to read it. Would it make sense to break it up into a few functions? For example, body of each conditional branch in a separate function or something along those lines. If we measured cyclomatic complexity as part of linting, this function's would be through the roof.

z3z1ma replied:
done

gouline commented on dbtmetabase/metabase.py:
```diff
@@ -404,14 +434,192 @@ def build_metadata_lookups(
+            os.path.expanduser(os.path.join(output_path, f"{output_name}.yml")), "w"
+        ) as docs:
+            yaml.dump(
+                {"version": 2, "exposures": captured_exposures},
```
Should the version be a constant somewhere? E.g. if dbt bumps it up in future.

z3z1ma replied:
adjusted to be constant at top of func

gouline commented on dbtmetabase/metabase.py:
```diff
@@ -404,14 +434,192 @@ def build_metadata_lookups(
 
         return table_lookup, field_lookup
 
+    def extract_exposures(
```
Would be nice to have header comments for these big important functions. How they work, inputs, outputs, etc.

z3z1ma replied:
done

gouline commented on dbtmetabase/metabase.py:
```diff
@@ -404,14 +434,192 @@ def build_metadata_lookups(
+                "source-table", model.get("table_id")
+            )
+            if source_table_id in self.table_map:
+                if isinstance(source_table_id, str) and source_table_id.startswith(
```
Funny formatting, maybe move `startswith` to the new line altogether.

z3z1ma replied:
simplified conditional to just use startswith on str(source_table_id) instead of nesting conditional check for str

z3z1ma commented on README.rst:
```diff
@@ -110,6 +112,63 @@ Check your Metabase instance by going into Settings > Admin > Data Model, you
+          name: Indiana Jones
+          email: user@example.com
+        depends_on:
+        - ref('orders')
```
That ones okay and nested properly

gouline replied:
See https://docs.getdbt.com/docs/building-a-dbt-project/exposures/#declaring-an-exposure

```
exposures:
  - name: weekly_jaffle_metrics
    type: dashboard
    maturity: high
    url: https://bi.tool/dashboards/1
    description: >
      Did someone say "exponential growth"?
    depends_on:
      - ref('fct_orders')
      - ref('dim_customers')
      - source('gsheets', 'goals')
```

z3z1ma replied:
Ah okay. Sorry, I guess itd be more accurate to say PyYaml thinks its accurate and valid. The above in the readme is taken directly from output of using the module. So it works, but see below issue. 

See https://github.com/yaml/pyyaml/issues/234 

I have implemented a solution which formats our exposures yaml output as dbt documents and will update in the readme to reflect.

gouline replied:
Ah right, the joys of YAML 😬

z3z1ma commented on dbtmetabase/metabase.py:
```diff
@@ -404,14 +434,192 @@ def build_metadata_lookups(
+    def extract_exposures(
+        self,
+        models: List[MetabaseModel],
+        output_path: str = "./",
```
thanks for that, make sense. adjusted

z3z1ma commented:
@gouline 

All polished up. I think its as good as can be now (and super readable). Thanks for the review and suggestions.

gouline reviewed (approved):
Awesome, mammoth effort! Thanks for persevering 🚀
